internal/backend/gs/gs.go

```diff
@@ -394,47 +368,36 @@ func (be *Backend) List(ctx context.Context, t restic.FileType, fn func(restic.F
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
-	listReq := be.service.Objects.List(be.bucketName).Context(ctx).Prefix(prefix).MaxResults(int64(be.listMaxItems))
-	for {
-		be.sem.GetToken()
-		obj, err := listReq.Do()
-		be.sem.ReleaseToken()
+	itr := be.bucket.Objects(ctx, &storage.Query{Prefix: prefix})
 
+	be.sem.GetToken()
```

Comment:
The semaphore should only protect the invocation of `itr.Next()` otherwise restic will deadlock when using `gcs.connections=1`. See the `List` method of `internal/backend/s3/s3.go` for a similar case.

Reply:
Thx! I moved it.